examples/step-66/step-66.cc

```diff
@@ -0,0 +1,1156 @@
+  // implementing the calculation of the cell contribution.
+
+  // TODO: As functor with operator() function?
+
+  template <int dim, int fe_degree>
+  class ResidualOperator
+  {
+  public:
+    ResidualOperator() = default;
+
+    void initialize(std::shared_ptr<const MatrixFree<dim, double>> data_in);
+
+    void apply(LinearAlgebra::distributed::Vector<double> &      dst,
+               const LinearAlgebra::distributed::Vector<double> &src) const;
+
+  private:
+    void
+    local_apply(const MatrixFree<dim, double> &                   data,
+                LinearAlgebra::distributed::Vector<double> &      dst,
+                const LinearAlgebra::distributed::Vector<double> &src,
+                const std::pair<unsigned int, unsigned int> &cell_range) const;
+
+    std::shared_ptr<const MatrixFree<dim, double>> data;
+  };
+
+
+
+  // The initialize function checks if given shared pointer of the MatrixFree
+  // object is not empty and pass it to the class varaiable.
+  template <int dim, int fe_degree>
+  void ResidualOperator<dim, fe_degree>::initialize(
+    std::shared_ptr<const MatrixFree<dim, double>> data_in)
+  {
+    Assert(data_in, ExcNotInitialized());
+
+    data = data_in;
+  }
+
+
+
+  // The main function evaluating the residual is just a wrapper around the
+  // cell_loop of the MatrixFree object.
+  template <int dim, int fe_degree>
+  void ResidualOperator<dim, fe_degree>::apply(
+    LinearAlgebra::distributed::Vector<double> &      dst,
+    const LinearAlgebra::distributed::Vector<double> &src) const
+  {
+    Assert(data, ExcNotInitialized());
+
+    data->cell_loop(
+      &ResidualOperator<dim, fe_degree>::local_apply, this, dst, src, true);
+  }
+
+
+
+  // The heart of the ResidualOperator is the local_apply function. The
+  // implementation is similar to the local_apply function of the above
+  // JacobianOperator. We setup a FEEvaluation object, gather and evaluate the
+  // values and gradients, submit the new values and gradients and finally
+  // integrate and distribute the local contributions to the global vector.
+  // Different to the Jacobian we need no additional table storing the values of
+  // the old Newton step, instead we can evaluate the nonlinearity on the fly,
+  // since we have to evaluate the residual in the input vector src,
+  // representing the last Newton step.
+  template <int dim, int fe_degree>
+  void ResidualOperator<dim, fe_degree>::local_apply(
+    const MatrixFree<dim> &                           data,
+    LinearAlgebra::distributed::Vector<double> &      dst,
+    const LinearAlgebra::distributed::Vector<double> &src,
+    const std::pair<unsigned int, unsigned int> &     cell_range) const
+  {
+    FEEvaluation<dim, fe_degree> phi(data);
+
+    for (unsigned int cell = cell_range.first; cell < cell_range.second; ++cell)
+      {
+        phi.reinit(cell);
+        phi.gather_evaluate(src, true, true);
```

Comment:
Same here, if you want inhomogeneous constraints, you need `read_dof_values_plain`.